dashboard-widgets.el

```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
+																			:action ,action
+																			:mouse-face 'highlight
+																			:follow-link "\C-m"
+																			:button-prefix ""
+																			:button-suffix ""
+																			:format "%[%t%]"
+																			,@rest))
+								 (setq max-line-length
+											 (max max-line-length (length (widget-value-value-get widget))))))
+						 ,list))
+		 max-line-length))
+
+(defmacro dashboard-insert-section (section-name list list-size shortcut action &rest widget-params)
+  "Add a section with SECTION-NAME and LIST of LIST-SIZE items to the dashboard.
+ACTION is theaction taken when the user activates the widget button.
+SHORTCUT is the keyboard shortcut used to access the section.
+WIDGET-PARAMS are passed to the \"widget-create\" function.
+Show EMPTY-LIST-TEXT if no items in list"
+	`(progn
+		 (dashboard-insert-heading ,section-name)
+		 (when-let ((max-line-length
+								 (dashboard-insert-section-list
+									,section-name
+									(dashboard-subseq ,list 0 list-size)
+									,action
+									,@widget-params)))
+			 (dashboard-insert-shortcut ,shortcut ,section-name)
+			 max-line-length)))
 
+;;
+;; Recentf
+;;
 (defun dashboard-insert-recents (list-size)
   "Add the list of LIST-SIZE items from recently edited files."
   (recentf-mode)
-  (when (dashboard-insert-recentf-list
+	(dashboard-insert-section
 	 "Recent Files:"
-	 (dashboard-subseq recentf-list 0 list-size))
-    (dashboard-insert-shortcut "r" "Recent Files:")))
-
+	 recentf-list
+	 list-size
+	 "r"
+	 `(lambda (&rest ignore) (find-file-existing ,el))
+	 (abbreviate-file-name el)))
 
 ;;
 ;; Bookmarks
 ;;
-(defun dashboard-insert-bookmark-list (list-display-name list)
-  "Render LIST-DISPLAY-NAME title and bookmarks items of LIST."
-  (when (car list)
-    (dashboard-insert-heading list-display-name)
-    (mapc (lambda (el)
-            (insert "\n    ")
-            (widget-create 'push-button
-                           :action `(lambda (&rest ignore) (bookmark-jump ,el))
-                           :mouse-face 'highlight
-                           :follow-link "\C-m"
-                           :button-prefix ""
-                           :button-suffix ""
-                           :format "%[%t%]"
-                           (let ((file (bookmark-get-filename el)))
-                             (if file
-                                 (format "%s - %s" el (abbreviate-file-name file))
-                               el))))
-          list)))
-
 (defun dashboard-insert-bookmarks (list-size)
   "Add the list of LIST-SIZE items of bookmarks."
   (require 'bookmark)
-  (when (dashboard-insert-bookmark-list
+  (dashboard-insert-section
 	 "Bookmarks:"
 	 (dashboard-subseq (bookmark-all-names)
-			   0 list-size))
-    (dashboard-insert-shortcut "m" "Bookmarks:")))
+																 0 list-size)
+	 list-size
+	 "m"
+	 `(lambda (&rest ignore) (bookmark-jump ,el))
+	 (let ((file (bookmark-get-filename el)))
+								 (if file
```

Comment:
Indentation is wrong here and the lines below

Reply:
Fixed!